source/Timer.cpp

```diff
@@ -0,0 +1,394 @@
+/* Timer.cpp
+Copyright (c) 2023 by Timothy Collett
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "Timer.h"
+
+#include "DataNode.h"
+#include "DataWriter.h"
+#include "GameData.h"
+#include "Logger.h"
+#include "Planet.h"
+#include "PlayerInfo.h"
+#include "Random.h"
+#include "Ship.h"
+#include "StellarObject.h"
+#include "UI.h"
+
+using namespace std;
+
+namespace {
+	string TriggerToText(Timer::TimerTrigger trigger)
+	{
+		switch(trigger)
+		{
+			case Timer::TimerTrigger::TIMEUP:
+				return "on timeup";
+			case Timer::TimerTrigger::RESET:
+				return "on reset";
+			default:
+				return "unknown trigger";
+		}
+	}
+}
+
+
+
+Timer::Timer(const DataNode &node)
+{
+	Load(node);
+}
+
+
+
+void Timer::Load(const DataNode &node)
+{
+	for(const DataNode &child : node)
+	{
+		if(child.Token(0) == "time" && child.Size() > 1)
+		{
+			waitTime = child.Value(1);
+			if(child.Size() > 2)
+				randomWaitTime = child.Value(2);
+		}
+		else if(child.Token(0) == "elapsed")
+			// This is only for saved timers; it is not intended for the data files.
+			timeElapsed = child.Value(1);
+		else if(child.Token(0) == "idle")
+		{
+			requireIdle = true;
+			// We square the max speed value here, so it can be conveniently
+			// compared to the flagship's squared velocity length below.
+			if(child.Size() > 1)
+			{
+				idleMaxSpeed = child.Value(1);
+				idleMaxSpeed *= idleMaxSpeed;
+			}
+		}
+		else if(child.Token(0) == "peaceful")
+			requirePeaceful = true;
+		else if(child.Token(0) == "optional")
+			optional = true;
+		else if(child.Token(0) == "system")
+		{
+			if(child.Size() > 1)
+				system = GameData::Systems().Get(child.Token(1));
+			else
+				systems.Load(child);
+		}
+		else if(child.Token(0) == "proximity")
+		{
+			if(child.Size() > 1)
+				proximityCenter = GameData::Planets().Get(child.Token(1));
+			else
+				proximityCenters.Load(child);
+		}
+		else if(child.Token(0) == "proximity settings" && child.Size() > 1)
+		{
+			proximity = child.Value(1);
+			if(child.Size() > 2)
+			{
+				if(child.Token(2) == "close")
+					closeTo = true;
+				else if(child.Token(2) == "far")
+					closeTo = false;
+			}
```

Comment:
closeTo defaults to true, so no need to check for "close" here.
```suggestion
			if(child.Size() > 2 && child.Token(2) == "far")
				closeTo = false;
```